indexing-service/src/main/java/org/apache/druid/indexing/common/task/batch/parallel/ParallelIndexPhaseRunner.java

```diff
@@ -0,0 +1,467 @@
+   */
+  abstract int getTotalNumSubTasks() throws IOException;
+
+  @Override
+  public TaskState run() throws Exception
+  {
+    if (getTotalNumSubTasks() == 0) {
+      LOG.warn("There's no input split to process");
+      return TaskState.SUCCESS;
+    }
+
+    final Iterator<SubTaskSpec<SubTaskType>> subTaskSpecIterator = subTaskSpecIterator();
+    final long taskStatusCheckingPeriod = tuningConfig.getTaskStatusCheckPeriodMs();
+
+    taskMonitor = new TaskMonitor<>(
+        Preconditions.checkNotNull(indexingServiceClient, "indexingServiceClient"),
+        tuningConfig.getMaxRetry(),
+        getTotalNumSubTasks()
+    );
+    TaskState state = TaskState.RUNNING;
+
+    taskMonitor.start(taskStatusCheckingPeriod);
+
+    try {
+      LOG.info("Submitting initial tasks");
+      // Submit initial tasks
+      while (isRunning() && subTaskSpecIterator.hasNext() && taskMonitor.getNumRunningTasks() < maxNumTasks) {
+        submitNewTask(taskMonitor, subTaskSpecIterator.next());
+      }
+
+      LOG.info("Waiting for subTasks to be completed");
+      while (isRunning()) {
+        final SubTaskCompleteEvent<SubTaskType> taskCompleteEvent = taskCompleteEvents.poll(
+            taskStatusCheckingPeriod,
+            TimeUnit.MILLISECONDS
+        );
+
+        if (taskCompleteEvent != null) {
+          final TaskState completeState = taskCompleteEvent.getLastState();
+          switch (completeState) {
+            case SUCCESS:
+              final TaskStatusPlus completeStatus = taskCompleteEvent.getLastStatus();
+              if (completeStatus == null) {
+                throw new ISE("Last status of complete task is missing!");
+              }
+              // Pushed segments of complete tasks are supposed to be already reported.
+              if (!reportsMap.containsKey(completeStatus.getId())) {
+                throw new ISE("Missing reports from task[%s]!", completeStatus.getId());
+              }
+
+              if (!subTaskSpecIterator.hasNext()) {
+                // We have no more subTasks to run
+                if (taskMonitor.getNumRunningTasks() == 0 && taskCompleteEvents.size() == 0) {
+                  subTaskScheduleAndMonitorStopped = true;
+                  if (taskMonitor.isSucceeded()) {
+                    // Succeeded
+                    state = TaskState.SUCCESS;
+                  } else {
+                    // Failed
+                    final SinglePhaseParallelIndexingProgress monitorStatus = taskMonitor.getProgress();
+                    throw new ISE(
+                        "Expected for [%d] tasks to succeed, but we got [%d] succeeded tasks and [%d] failed tasks",
+                        monitorStatus.getExpectedSucceeded(),
+                        monitorStatus.getSucceeded(),
+                        monitorStatus.getFailed()
+                    );
+                  }
+                }
+              } else if (taskMonitor.getNumRunningTasks() < maxNumTasks) {
```

Comment:
What would trigger this if case? It looks like the `subTaskSpecIterator` would be exhausted before entering the `while (isRunning())` loop and all the subtasks should've been submitted already

Reply:
This can happen if the number of sub tasks to execute is larger than `maxNumTasks` which is "the max number of sub tasks which can be executed concurrently at the same time". Renamed it to `maxNumConcurrentSubTasks` and added a javadoc.